src/rpc/auxcache.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2025 The Dogecoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "rpc/auxcache.h"
+
+#include "script/standard.h"  // for CScriptID
+#include "primitives/block.h" // for CBlock
+#include "sync.h"             // for LOCK
+#include "uint256.h"          // for uint256
+#include "utilmemory.h"       // for MakeUnique
+#include "utiltime.h"         // for GetTimeMicros
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <cstdint>            // for uint64_t
+#include <memory>             // for std::unique_ptr, std::shared_ptr
+#include <mutex>              // for std::mutex
+#include <tuple>              // for std::tuple
+
+namespace {
+
+//! Type that is stored in the cache
+struct AuxCacheItem {
+  int64_t cache_time;
+  CScriptID scriptId;
+  uint256 hash;
+  std::shared_ptr<CBlock> pblock;
+};
+
+// The ByScriptId index sorts by CScriptID and item age
+struct ByScriptId {};
+using ByScriptIdView = std::tuple<const CScriptID&, const int64_t>;
+struct ByScriptIdExtractor
+{
+    using result_type = ByScriptIdView;
+    result_type operator()(const AuxCacheItem& item) const
+    {
+        // calculate age
+        const int64_t age = GetMockableTimeMicros() - item.cache_time;
+        return ByScriptIdView{item.scriptId, age};
+    }
+};
+
+// The ByBlockHash index sorts by blockhash (uint256)
+struct ByBlockHash {};
+struct ByBlockHashExtractor
+{
+    using result_type = uint256;
+    result_type operator()(const AuxCacheItem& item) const
+    {
+        return item.hash;
+    }
+};
+
+/** Data type for the main data structure (AuxCacheItem objects with ByScriptID/ByBlockHash indexes). */
+using AuxCacheIndex = boost::multi_index_container<
+    AuxCacheItem,
+    boost::multi_index::indexed_by<
+        boost::multi_index::ordered_non_unique<boost::multi_index::tag<ByScriptId>, ByScriptIdExtractor>,
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByBlockHash>, ByBlockHashExtractor>
+    >
```

Comment:
It would be nice if C++ someday offered an STL replacement for this, but Boost it is for now!